presto-main/src/main/java/com/facebook/presto/catalogserver/CatalogServer.java

```diff
@@ -24,69 +24,116 @@
         this.transactionManager = requireNonNull(transactionManager, "transactionManager is null");
         this.objectMapper = requireNonNull(objectMapper, "handleResolver is null");
+
+        CacheBuilder<Object, Object> cacheBuilder = CacheBuilder.newBuilder()
+                .maximumSize(CACHE_MAXIMUM_SIZE)
+                .expireAfterWrite(CACHE_EXPIRES_AFTER_WRITE_MILLIS.toMillis(), MILLISECONDS);
+
+        this.catalogExistsCache = cacheBuilder.build(CacheLoader.from(this::loadCatalogExists));
+        this.schemaExistsCache = cacheBuilder.build(CacheLoader.from(this::loadSchemaExists));
+        this.listSchemaNamesCache = cacheBuilder.build(CacheLoader.from(this::loadListSchemaNames));
+        this.getTableHandleCache = cacheBuilder.build(CacheLoader.from(this::loadGetTableHandle));
+        this.listTablesCache = cacheBuilder.build(CacheLoader.from(this::loadListTables));
+        this.listViewsCache = cacheBuilder.build(CacheLoader.from(this::loadListViews));
+        this.getViewsCache = cacheBuilder.build(CacheLoader.from(this::loadGetViews));
+        this.getViewCache = cacheBuilder.build(CacheLoader.from(this::loadGetView));
+        this.getMaterializedViewCache = cacheBuilder.build(CacheLoader.from(this::loadGetMaterializedView));
+        this.getReferencedMaterializedViewsCache = cacheBuilder.build(CacheLoader.from(this::loadGetReferencedMaterializedViews));
     }
 
+    /*
+        Metadata Manager Methods
+     */
+
     @ThriftMethod
     public boolean schemaExists(TransactionInfo transactionInfo, SessionRepresentation session, CatalogSchemaName schema)
     {
         transactionManager.tryRegisterTransaction(transactionInfo);
-        return metadataProvider.schemaExists(session.toSession(sessionPropertyManager), schema);
+        return get(schemaExistsCache, new CacheKey<>(session, schema));
     }
 
     @ThriftMethod
     public boolean catalogExists(TransactionInfo transactionInfo, SessionRepresentation session, String catalogName)
     {
         transactionManager.tryRegisterTransaction(transactionInfo);
-        return metadataProvider.catalogExists(session.toSession(sessionPropertyManager), catalogName);
+        return get(catalogExistsCache, new CacheKey(session, catalogName));
     }
 
     @ThriftMethod
     public String listSchemaNames(TransactionInfo transactionInfo, SessionRepresentation session, String catalogName)
     {
         transactionManager.tryRegisterTransaction(transactionInfo);
-        List<String> schemaNames = metadataProvider.listSchemaNames(session.toSession(sessionPropertyManager), catalogName);
-        if (!schemaNames.isEmpty()) {
-            return writeValueAsString(schemaNames, objectMapper);
-        }
-        return EMPTY_STRING;
+        List<String> schemaNames = get(listSchemaNamesCache, new CacheKey(session, catalogName));
+        return schemaNames.isEmpty()
+                ? EMPTY_STRING
+                : writeValueAsString(schemaNames, objectMapper);
```

Comment:
Didn't read the code carefully before. But why do we need do serde by ourselves into strings? Can we use the thrift objects? For example `List<String>` is already serializable in thrift. Same for many other objects in the class like `QualifiedObjectName`.

@ajaygeorge might be able to advise what would be a good way to define the protocol here.